### PR TITLE
[Serialization] Ignore conformance failures when allowing errors

### PR DESCRIPTION
**Pull request by bnbarham**
Allowing errors is attempting to create a module regardless of any
potential failures, so do our best to continue rather than crash (even
if a corresponding build would then crash).

### CODE REVIEW

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
I need to add a ObjectiveC import to properly test the `finishNormalConformance` case as well, but I thought I'd put up what I had to see if this is something we actually want before doing that. This is just what I had from reducing and attempting to fix a crasher.

File: lib/Serialization/Deserialization.cpp [hunk 1]
```diff
@@ -4473,10 +4473,12 @@ llvm::Error DeclDeserializer::deserializeDeclCommon() {
 
         Expected<Type> deserialized = MF.getTypeChecked(typeID);
         if (!deserialized) {
-          if (deserialized.errorIsA<XRefNonLoadedModuleError>()) {
+          if (deserialized.errorIsA<XRefNonLoadedModuleError>() ||
+              MF.allowCompilerErrors()) {
```
Comment (xymus) [inline lib/Serialization/Deserialization.cpp:4477]: Yea, it's a good idea to also cover the `XRefNonLoadedModuleError` cases for `MF.allowCompilerErrors`.

File: lib/Serialization/Deserialization.cpp [hunk 2]
```diff
@@ -6359,8 +6367,17 @@ void ModuleFile::finishNormalConformance(NormalProtocolConformance *conformance,
...
     }
-    while (conformanceCount--)
-      reqConformances.push_back(readConformance(DeclTypeCursor));
+    while (conformanceCount--) {
+      auto conformance = readConformanceChecked(DeclTypeCursor);
+      if (conformance) {
+        reqConformances.push_back(conformance.get());
+      } else if (getContext().LangOpts.EnableDeserializationRecovery) {
+        consumeError(conformance.takeError());
+        reqConformances.push_back(ProtocolConformanceRef::forInvalid());
```
Comment (xymus) [inline lib/Serialization/Deserialization.cpp:6376]: An extra recovery here makes sense with the one above.

What is the error in normal compilation when this invalid conformance is passed down again?

Comment (bnbarham) [reply lib/Serialization/Deserialization.cpp:6376]: I'll add the ObjC test and have a look :)

**Comment by bnbarham** [comment]
@swift-ci please test

File: lib/Serialization/Deserialization.cpp (hunk 2 above)
Comment (bnbarham) [reply lib/Serialization/Deserialization.cpp:6398]: Okay, so turns out this needs some protocol and associated type fun to reproduce, not Objc. But I've got both now anyway! See the updated test.

I couldn't hit just this case without it crashing in the witness substitution (when not allowing errors). When allowing errors this ends up showing up as a conformance diagnostic:
```
error: referencing static method 'thing' on 'ReqProto' requires that 'EnumB.Value' conform to 'ProtoA'
```

Which is about as helpful as the missing type diagnostics:
```
error: cannot find type 'ProtoB' in scope
```

**Comment by bnbarham** [comment]
@swift-ci please test

**Review by xymus — APPROVED** [review]

**Comment by bnbarham** [comment]
@swift-ci please test

**Pull request merged**